lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java

```diff
@@ -813,8 +881,28 @@ public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
 
       if (docIdSet == null) {
         if (policy.shouldCache(in.getQuery())) {
-          docIdSet = cache(context);
-          putIfAbsent(in.getQuery(), docIdSet, cacheHelper);
+          boolean performSynchronousCaching = !(executor != null);
+          // If asynchronous caching is requested, perform the same and return
+          // the uncached iterator
+          if (!performSynchronousCaching) {
+            try {
+              cacheAsynchronously(context, cacheHelper);
+            } catch (RejectedExecutionException e) {
```

Comment:
Hmm, can you shrink-wrap this exception handling?  I.e., handle it down in `cacheAsynchronously` where we ask the executor to execute the task, instead of up here?

Maybe then return a `boolean` from `cacheAsynchronously` to know (here) if caller should then cache synchronously ...